### DISCUSSION TITLE
InDesignでの数式画像のベースライン統一方法

### DISCUSSION THREAD

**Discussion by kn1cht** [kmuto/review#1830] [General]
#1829 で、InDesign XML出力でも数式を画像化するimgmathが使えるようになりました。

ただし、生成された数式画像は**中身の文字の高さによって異なるサイズ**になります。そのため、単純にインラインオブジェクトとして下揃えなどで配置してもベースラインがバラバラになってしまいます。

![image](https://user-images.githubusercontent.com/16277200/178135962-0f49bf82-30e2-416c-829a-bb564ac90af2.png)

本Discussionで、数式画像のベースライン問題の対処法について情報交換したく思います。

思い浮かぶ対処法とその分類を示しておきます。

- A: 数式画像の縦方向サイズを統一する
    - A-1: 全画像で、縦方向サイズとベースラインの上下位置を統一する
    - A-2: 全画像で、ベースラインから下端までの距離を統一する
        - LaTeX2HTMLでは、[部分的にこの方針が採られているようです](http://takeno.iee.niit.ac.jp/~shige/TeX/latex2html/l2h-notice.html#20050731-2)
- B: 数式画像の位置調整のための情報をInDesignに渡す
    - B-1: 数式画像自体に、ベースラインの位置情報を付加する
    - B-2: XMLタグの属性に、ベースラインの位置情報を付加する
    - B-3: 数式画像とは別のファイルに、ベースラインの位置情報を付加する
        - kmuto様のInDesign Hacking with JavaScriptに含まれる[`dialog2TweakMathImages.jsx`](https://kmuto.jp/indesign/scripts/dialogs/dialog2TweakMathImages.jsx)はこの方針でしょうか？
- C: そもそもRe:VIEW段階では画像化しない
    - C-1: 文字スタイルで数式を表現する
        - [desperatex](https://github.com/kmuto/desperatex)
    - C-2: InDesign側で数式画像化のスクリプトを実行する
        - [inDesignにTeXの数式を貼り込むスクリプト「インてふてふ」](http://sysys.blog.shinobi.jp/Entry/18/)

**Comment by kn1cht** [comment]
思いついたA-2「全画像で、ベースラインから下端までの距離を統一する」の実現方法があるので書き残しておきます（ただし、かなり黒魔術的です…）

1. imgmathで数式画像を切り抜くのに使われている[`pdfcrop`](https://github.com/ho-tex/pdfcrop)は、描画がある部分以外を切り抜いてしまいます。
2. そこで、ベースラインから一定距離に何らかの描画をして、下端からベースラインまでの距離を揃えます。
    - 例えば、任意の大きさの黒い四角を生成する`\rule`コマンドは、オプションでベースラインからの距離を指定できます。
    - IDGXMLBuilderで `math_str = '$\\rule[-5pt]{0.01pt}{0.01pt}' + str + '$'` のようにすると下揃えでベースラインが揃います（参考画像では、見やすいよう四角を大きめにしました）。
    - ![image](https://user-images.githubusercontent.com/16277200/178137441-b841ef0a-00e4-4c77-8734-2186214e535d.png)
3. このまま`pdfcrop`に掛けると、画像下の四角が見えてしまいます。`pdfcrop`では`--margins`オプションで負のmarginを指定できるので、下だけ数pt余分に切り抜きます。
    - `pdfcrop_cmd: "pdfcrop --hires --margins '0 0 0 -2' %i %o"`
    - ![image](https://user-images.githubusercontent.com/16277200/178137617-5cd99236-49c3-417c-bc05-813263cd0b6a.png)
4. これで、数式画像同士ではサイズとベースライン位置が合ったので、InDesign側でサイズや上下位置を一括で調節すればOKです

**Comment by kmuto** [comment]
なるほど、アプローチ提案ありがとうございます。 :smiley: 
インライン範囲だとこれでもいけそうなのですが、行送り内に収まらない形だと突き抜けて詰みそうに思いました(実際、商業書籍だとインライン数式形状は絶対NGでディスプレイ数式にして前後行送りを調整せよという指定のほうが多かったり)。

dialog2TweakMathImages.jsxが参照する情報ファイルの元ネタは社内ツールで出せない状態なのですが、アプローチとしては以下のとおりです。

1. 数式の先頭にベースラインに合わせて横線を引いた数式画像を作る。
2. 全体の高さとベースラインの位置を取得する。具体的には1.をビットマップ変換して左側で色ピクセルが見つかるまで縦方向で探す。これでベースラインからのシフト率が出せる。
3. 事前に算出しておいた基準となる文字の高さと、2.の高さの比で拡縮率を計算する。
4. 横線なしバージョンの本番数式画像を作る。
5. ファイル名、ベースラインからのシフト率と拡縮率を列挙する。
6. スクリプトで該当ファイル名に対して処理を実行。

4.のところは1.からcropのマージン指定で作れるかもしれませんが、全体のスケールを大きくして作りたいといったとき(「i」「j」などの細い文字を小さな状態でpdfcropすると壊れたデータになります…)にマージン値がそれに応じて変動するので、結局作り直したほうが早いというのが自分の結論でした。
業務用のツールのほうだと、ファイル名もハッシュ値ではなく数式の内容に即した名前(y=x^2.pdfとかLS_x=sigmax_i.pdfとか)にしていたりもします(制作途中から組み直しでなく数式ファイルを手で差し替えていくことがよくあり、DTPオペレータのファイル名探しの負担を少しでも軽減するため)。

**Comment by kn1cht** [comment]
貴重な情報ありがとうございます！
ディスプレイ数式もあるとなると、画像の縦サイズを揃えるアプローチでは限界がありそうですね…
